homeassistant/components/nibe_heatpump/__init__.py

```diff
@@ -0,0 +1,242 @@
+"""The Nibe Heat Pump integration."""
+from __future__ import annotations
+
+import asyncio
+from datetime import timedelta
+import logging
+
+from nibe.coil import Coil
+from nibe.connection import Connection
+from nibe.connection.nibegw import NibeGW
+from nibe.exceptions import CoilNotFoundException, CoilReadException
+from nibe.heatpump import HeatPump, Model
+
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.const import CONF_IP_ADDRESS, CONF_MODEL, Platform
+from homeassistant.core import CALLBACK_TYPE, HomeAssistant
+from homeassistant.exceptions import HomeAssistantError
+from homeassistant.helpers import device_registry as dr
+from homeassistant.helpers.entity import DeviceInfo, async_generate_entity_id
+from homeassistant.helpers.update_coordinator import (
+    CoordinatorEntity,
+    DataUpdateCoordinator,
+)
+
+from .const import (
+    CONF_CONNECTION_TYPE,
+    CONF_CONNECTION_TYPE_NIBEGW,
+    CONF_LISTENING_PORT,
+    CONF_REMOTE_READ_PORT,
+    CONF_REMOTE_WRITE_PORT,
+    CONF_WORD_SWAP,
+    DOMAIN,
+)
+from .utils import TooManyTriesException, retry
+
+_LOGGER = logging.getLogger(__name__)
+
+PLATFORMS: list[Platform] = [Platform.SENSOR]
+
+
+async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
+    """Set up Nibe Heat Pump from a config entry."""
+
+    heatpump = HeatPump(Model[entry.data[CONF_MODEL]])
+    heatpump.word_swap = entry.data[CONF_WORD_SWAP]
+    heatpump.initialize()
+
+    connection_type = entry.data[CONF_CONNECTION_TYPE]
+
+    if connection_type == CONF_CONNECTION_TYPE_NIBEGW:
+        connection = NibeGW(
+            heatpump,
+            entry.data[CONF_IP_ADDRESS],
+            entry.data[CONF_REMOTE_READ_PORT],
+            entry.data[CONF_REMOTE_WRITE_PORT],
+            listening_port=entry.data[CONF_LISTENING_PORT],
+        )
+    else:
+        raise HomeAssistantError(f"Connection type {connection_type} is not supported.")
+
+    await connection.start()
+    try:
```

Comment:
This try/except block doesn't feel right:

1. We're including a lot inside of it, which doesn't explicitly indicate where an exception might occur (are we responding to a coordinator update error? an error with the device registry? something else?)
2. We're broadly catching every exception – can we be more specific than that?

Reply:
We are just making sure to clean up on unknown exceptions, so we don't leave a lingering connection.

Reply:
You could do

```
try:
        await coordinator.async_config_entry_first_refresh()
except ConfigEntryNotReady:
        await connection.stop()
        raise

```

Reply:
Well, the point of the code is to catch unknown errors and not leak connections.

Reply:
I think you would still be ok as `async_config_entry_first_refresh` has a broad exception catch under the hood so it will only return or raise `ConfigEntryNotReady`

Reply:
Since there seem to be a preference for tight exception handling, I've done as suggested even thou it's now less robust against unexpected failures.